// @zip.js/zip.js/lib/core/streams/codec-stream.very.original.js
import { ERR_INVALID_PASSWORD, ERR_INVALID_SIGNATURE, InflateStream, DeflateStream } from "./zip-entry-stream.js";
const MESSAGE_EVENT_TYPE = "message";
const MESSAGE_START = "start";
const MESSAGE_PULL = "pull";
const MESSAGE_DATA = "data";
const MESSAGE_ACK_DATA = "ack";
const MESSAGE_CLOSE = "close";
const CODEC_DEFLATE = "deflate";
const CODEC_INFLATE = "inflate";
export { CODEC_DEFLATE, CODEC_INFLATE, MESSAGE_EVENT_TYPE, MESSAGE_START, MESSAGE_PULL, MESSAGE_DATA, MESSAGE_ACK_DATA, MESSAGE_CLOSE, ERR_INVALID_PASSWORD, ERR_INVALID_SIGNATURE, CodecStream };
class CodecStream extends TransformStream {
    constructor(options, config) {
        super({});
        const codec = this;
        const { codecType } = options;
        let Stream;
        if (codecType.startsWith(CODEC_DEFLATE)) {
            Stream = DeflateStream;
        }
        else if (codecType.startsWith(CODEC_INFLATE)) {
            Stream = InflateStream;
        }
        let size = 0;
        const stream = new Stream(options, config);
        const readable = super.readable;
        const transformStream = new TransformStream({
            transform(chunk, controller) {
                if (chunk && chunk.length) {
                    size += chunk.length;
                    controller.enqueue(chunk);
                }
            },
            flush() {
                const { signature } = stream;
                Object.assign(codec, {
                    signature,
                    size
                });
            }
        });
        Object.defineProperty(codec, "readable", {
            get() {
                return readable.pipeThrough(stream).pipeThrough(transformStream);
            }
        });
    }
}
