// @zip.js/zip.js/lib/core/streams/zip-crypto-stream.very.original.js
import { Crc32 } from "./codecs/crc32.js";
import { ERR_INVALID_PASSWORD, getRandomValues } from "./common-crypto.js";
const HEADER_LENGTH = 12;
class ZipCryptoDecryptionStream extends TransformStream {
    constructor({ password, passwordVerification }) {
        super({
            start() {
                Object.assign(this, {
                    password,
                    passwordVerification
                });
                createKeys(this, password);
            },
            transform(chunk, controller) {
                const zipCrypto = this;
                if (zipCrypto.password) {
                    const decryptedHeader = decrypt(zipCrypto, chunk.subarray(0, HEADER_LENGTH));
                    zipCrypto.password = null;
                    if (decryptedHeader[HEADER_LENGTH - 1] != zipCrypto.passwordVerification) {
                        throw new Error(ERR_INVALID_PASSWORD);
                    }
                    chunk = chunk.subarray(HEADER_LENGTH);
                }
                controller.enqueue(decrypt(zipCrypto, chunk));
            }
        });
    }
}
class ZipCryptoEncryptionStream extends TransformStream {
    constructor({ password, passwordVerification }) {
        super({
            start() {
                Object.assign(this, {
                    password,
                    passwordVerification
                });
                createKeys(this, password);
            },
            transform(chunk, controller) {
                const zipCrypto = this;
                let output;
                let offset;
                if (zipCrypto.password) {
                    zipCrypto.password = null;
                    const header = getRandomValues(new Uint8Array(HEADER_LENGTH));
                    header[HEADER_LENGTH - 1] = zipCrypto.passwordVerification;
                    output = new Uint8Array(chunk.length + header.length);
                    output.set(encrypt(zipCrypto, header), 0);
                    offset = HEADER_LENGTH;
                }
                else {
                    output = new Uint8Array(chunk.length);
                    offset = 0;
                }
                output.set(encrypt(zipCrypto, chunk), offset);
                controller.enqueue(output);
            }
        });
    }
}
export { ZipCryptoDecryptionStream, ZipCryptoEncryptionStream, ERR_INVALID_PASSWORD };
function decrypt(target, input) {
    const output = new Uint8Array(input.length);
    for (let index = 0; index < input.length; index++) {
        output[index] = getByte(target) ^ input[index];
        updateKeys(target, output[index]);
    }
    return output;
}
function encrypt(target, input) {
    const output = new Uint8Array(input.length);
    for (let index = 0; index < input.length; index++) {
        output[index] = getByte(target) ^ input[index];
        updateKeys(target, input[index]);
    }
    return output;
}
function createKeys(target, password) {
    const keys = [0x12345678, 0x23456789, 0x34567890];
    Object.assign(target, {
        keys,
        crcKey0: new Crc32(keys[0]),
        crcKey2: new Crc32(keys[2]),
    });
    for (let index = 0; index < password.length; index++) {
        updateKeys(target, password.charCodeAt(index));
    }
}
function updateKeys(target, byte) {
    let [key0, key1, key2] = target.keys;
    target.crcKey0.append([byte]);
    key0 = ~target.crcKey0.get();
    key1 = getInt32(Math.imul(getInt32(key1 + getInt8(key0)), 134775813) + 1);
    target.crcKey2.append([key1 >>> 24]);
    key2 = ~target.crcKey2.get();
    target.keys = [key0, key1, key2];
}
function getByte(target) {
    const temp = target.keys[2] | 2;
    return getInt8(Math.imul(temp, (temp ^ 1)) >>> 8);
}
function getInt8(number) {
    return number & 0xFF;
}
function getInt32(number) {
    return number & 0xFFFFFFFF;
}
