// @zip.js/zip.js/lib/core/streams/zip-entry-stream.very.original.js
import { Crc32Stream } from "./crc32-stream.js";
import { AESEncryptionStream, AESDecryptionStream } from "./aes-crypto-stream.js";
import { ZipCryptoEncryptionStream, ZipCryptoDecryptionStream } from "./zip-crypto-stream.js";
import { ERR_INVALID_PASSWORD, ERR_INVALID_SIGNATURE } from "./common-crypto.js";
const COMPRESSION_FORMAT = "deflate-raw";
class DeflateStream extends TransformStream {
    constructor(options, { chunkSize, CompressionStream, CompressionStreamNative }) {
        super({});
        const { compressed, encrypted, useCompressionStream, zipCrypto, signed, level } = options;
        const stream = this;
        let crc32Stream, encryptionStream;
        let readable = filterEmptyChunks(super.readable);
        if ((!encrypted || zipCrypto) && signed) {
            [readable, crc32Stream] = readable.tee();
            crc32Stream = pipeThrough(crc32Stream, new Crc32Stream());
        }
        if (compressed) {
            readable = pipeThroughCommpressionStream(readable, useCompressionStream, { level, chunkSize }, CompressionStreamNative, CompressionStream);
        }
        if (encrypted) {
            if (zipCrypto) {
                readable = pipeThrough(readable, new ZipCryptoEncryptionStream(options));
            }
            else {
                encryptionStream = new AESEncryptionStream(options);
                readable = pipeThrough(readable, encryptionStream);
            }
        }
        setReadable(stream, readable, async () => {
            let signature;
            if (encrypted && !zipCrypto) {
                signature = encryptionStream.signature;
            }
            if ((!encrypted || zipCrypto) && signed) {
                signature = await crc32Stream.getReader().read();
                signature = new DataView(signature.value.buffer).getUint32(0);
            }
            stream.signature = signature;
        });
    }
}
class InflateStream extends TransformStream {
    constructor(options, { chunkSize, DecompressionStream, DecompressionStreamNative }) {
        super({});
        const { zipCrypto, encrypted, signed, signature, compressed, useCompressionStream } = options;
        let crc32Stream, decryptionStream;
        let readable = filterEmptyChunks(super.readable);
        if (encrypted) {
            if (zipCrypto) {
                readable = pipeThrough(readable, new ZipCryptoDecryptionStream(options));
            }
            else {
                decryptionStream = new AESDecryptionStream(options);
                readable = pipeThrough(readable, decryptionStream);
            }
        }
        if (compressed) {
            readable = pipeThroughCommpressionStream(readable, useCompressionStream, { chunkSize }, DecompressionStreamNative, DecompressionStream);
        }
        if ((!encrypted || zipCrypto) && signed) {
            [readable, crc32Stream] = readable.tee();
            crc32Stream = pipeThrough(crc32Stream, new Crc32Stream());
        }
        setReadable(this, readable, async () => {
            if ((!encrypted || zipCrypto) && signed) {
                const streamSignature = await crc32Stream.getReader().read();
                const dataViewSignature = new DataView(streamSignature.value.buffer);
                if (signature != dataViewSignature.getUint32(0, false)) {
                    throw new Error(ERR_INVALID_SIGNATURE);
                }
            }
        });
    }
}
export { DeflateStream, InflateStream, ERR_INVALID_PASSWORD, ERR_INVALID_SIGNATURE };
function filterEmptyChunks(readable) {
    return pipeThrough(readable, new TransformStream({
        transform(chunk, controller) {
            if (chunk && chunk.length) {
                controller.enqueue(chunk);
            }
        }
    }));
}
function setReadable(stream, readable, flush) {
    readable = pipeThrough(readable, new TransformStream({ flush }));
    Object.defineProperty(stream, "readable", {
        get() {
            return readable;
        }
    });
}
function pipeThroughCommpressionStream(readable, useCompressionStream, options, CodecStreamNative, CodecStream) {
    try {
        const CompressionStream = useCompressionStream && CodecStreamNative ? CodecStreamNative : CodecStream;
        readable = pipeThrough(readable, new CompressionStream(COMPRESSION_FORMAT, options));
    }
    catch (error) {
        if (useCompressionStream) {
            readable = pipeThrough(readable, new CodecStream(COMPRESSION_FORMAT, options));
        }
        else {
            throw error;
        }
    }
    return readable;
}
function pipeThrough(readable, transformStream) {
    return readable.pipeThrough(transformStream);
}
