// @zip.js/zip.js/lib/core/codec-pool.very.original.js
import { UNDEFINED_VALUE } from "./constants.js";
import { CODEC_INFLATE, CODEC_DEFLATE, ERR_INVALID_SIGNATURE, ERR_INVALID_PASSWORD } from "./streams/codec-stream.js";
import { CodecWorker } from "./codec-worker.js";
let pool = [];
const pendingRequests = [];
export { runWorker, terminateWorkers, CODEC_DEFLATE, CODEC_INFLATE, ERR_INVALID_SIGNATURE, ERR_INVALID_PASSWORD };
let indexWorker = 0;
async function runWorker(stream, workerOptions) {
    const { options, config } = workerOptions;
    const { transferStreams, useWebWorkers, useCompressionStream, codecType, compressed, signed, encrypted } = options;
    const { workerScripts, maxWorkers, terminateWorkerTimeout } = config;
    workerOptions.transferStreams = transferStreams || transferStreams === UNDEFINED_VALUE;
    const streamCopy = !compressed && !signed && !encrypted && !workerOptions.transferStreams;
    workerOptions.useWebWorkers = !streamCopy && (useWebWorkers || (useWebWorkers === UNDEFINED_VALUE && config.useWebWorkers));
    workerOptions.scripts = workerOptions.useWebWorkers && workerScripts ? workerScripts[codecType] : [];
    options.useCompressionStream = useCompressionStream || (useCompressionStream === UNDEFINED_VALUE && config.useCompressionStream);
    let worker;
    const workerData = pool.find(workerData => !workerData.busy);
    if (workerData) {
        clearTerminateTimeout(workerData);
        worker = new CodecWorker(workerData, stream, workerOptions, onTaskFinished);
    }
    else if (pool.length < maxWorkers) {
        const workerData = { indexWorker };
        indexWorker++;
        pool.push(workerData);
        worker = new CodecWorker(workerData, stream, workerOptions, onTaskFinished);
    }
    else {
        worker = await new Promise(resolve => pendingRequests.push({ resolve, stream, workerOptions }));
    }
    return worker.run();
    function onTaskFinished(workerData) {
        if (pendingRequests.length) {
            const [{ resolve, stream, workerOptions }] = pendingRequests.splice(0, 1);
            resolve(new CodecWorker(workerData, stream, workerOptions, onTaskFinished));
        }
        else if (workerData.worker) {
            clearTerminateTimeout(workerData);
            if (Number.isFinite(terminateWorkerTimeout) && terminateWorkerTimeout >= 0) {
                workerData.terminateTimeout = setTimeout(() => {
                    pool = pool.filter(data => data != workerData);
                    workerData.terminate();
                }, terminateWorkerTimeout);
            }
        }
        else {
            pool = pool.filter(data => data != workerData);
        }
    }
}
function clearTerminateTimeout(workerData) {
    const { terminateTimeout } = workerData;
    if (terminateTimeout) {
        clearTimeout(terminateTimeout);
        workerData.terminateTimeout = null;
    }
}
function terminateWorkers() {
    pool.forEach(workerData => {
        clearTerminateTimeout(workerData);
        workerData.terminate();
    });
}
