// @zip.js/zip.js/lib/core/codec-worker.original.js
import { UNDEFINED_TYPE, FUNCTION_TYPE } from "./constants.js";
import { CodecStream, MESSAGE_EVENT_TYPE, MESSAGE_START, MESSAGE_PULL, MESSAGE_DATA, MESSAGE_ACK_DATA, MESSAGE_CLOSE } from "./streams/codec-stream.js";
const WEB_WORKERS_SUPPORTED = typeof Worker != UNDEFINED_TYPE;
export { CodecWorker };
class CodecWorker {
    constructor(workerData, { readable, writable }, { options, config, streamOptions, useWebWorkers, transferStreams, scripts }, onTaskFinished) {
        const { signal } = streamOptions;
        Object.assign(workerData, {
            busy: true,
            readable: readable.pipeThrough(new ProgressWatcherStream(readable, streamOptions, config), { signal }),
            writable,
            options: Object.assign({}, options),
            scripts,
            transferStreams,
            terminate() {
                const { worker, busy } = workerData;
                if (worker && !busy) {
                    worker.terminate();
                    workerData.interface = null;
                }
            },
            onTaskFinished() {
                workerData.busy = false;
                onTaskFinished(workerData);
            }
        });
        return (useWebWorkers && WEB_WORKERS_SUPPORTED ? createWebWorkerInterface : createWorkerInterface)(workerData, config);
    }
}
class ProgressWatcherStream extends TransformStream {
    constructor(readableSource, { onstart, onprogress, size, onend }, { chunkSize }) {
        let chunkOffset = 0;
        super({
            start() {
                if (onstart) {
                    callHandler(onstart, size);
                }
            },
            async transform(chunk, controller) {
                chunkOffset += chunk.length;
                if (onprogress) {
                    await callHandler(onprogress, chunkOffset, size);
                }
                controller.enqueue(chunk);
            },
            flush() {
                readableSource.size = chunkOffset;
                if (onend) {
                    callHandler(onend, chunkOffset);
                }
            }
        }, { highWaterMark: 1, size: () => chunkSize });
    }
}
async function callHandler(handler, ...parameters) {
    try {
        await handler(...parameters);
    }
    catch (_error) {
    }
}
function createWorkerInterface(workerData, config) {
    return {
        run: () => runWorker(workerData, config)
    };
}
function createWebWorkerInterface(workerData, { baseURL, chunkSize }) {
    if (!workerData.interface) {
        Object.assign(workerData, {
            worker: getWebWorker(workerData.scripts[0], baseURL, workerData),
            interface: {
                run: () => runWebWorker(workerData, { chunkSize })
            }
        });
    }
    return workerData.interface;
}
async function runWorker({ options, readable, writable, onTaskFinished }, config) {
    const codecStream = new CodecStream(options, config);
    try {
        await readable.pipeThrough(codecStream).pipeTo(writable, { preventClose: true, preventAbort: true });
        const { signature, size } = codecStream;
        return {
            signature,
            size
        };
    }
    finally {
        onTaskFinished();
    }
}
async function runWebWorker(workerData, config) {
    let resolveResult, rejectResult;
    const result = new Promise((resolve, reject) => {
        resolveResult = resolve;
        rejectResult = reject;
    });
    Object.assign(workerData, {
        reader: null,
        writer: null,
        resolveResult,
        rejectResult,
        result
    });
    const { readable, options, scripts } = workerData;
    const { writable, closed } = watchClosedStream(workerData.writable);
    const streamsTransferred = sendMessage({
        type: MESSAGE_START,
        scripts: scripts.slice(1),
        options,
        config,
        readable,
        writable
    }, workerData);
    if (!streamsTransferred) {
        Object.assign(workerData, {
            reader: readable.getReader(),
            writer: writable.getWriter()
        });
    }
    const resultValue = await result;
    try {
        await writable.close();
    }
    catch (_error) {
    }
    await closed;
    return resultValue;
}
function watchClosedStream(writableSource) {
    const writer = writableSource.getWriter();
    let resolveStreamClosed;
    const closed = new Promise(resolve => resolveStreamClosed = resolve);
    const writable = new WritableStream({
        async write(chunk) {
            await writer.ready;
            await writer.write(chunk);
        },
        close() {
            writer.releaseLock();
            resolveStreamClosed();
        },
        abort(reason) {
            return writer.abort(reason);
        }
    });
    return { writable, closed };
}
let classicWorkersSupported = true;
let transferStreamsSupported = true;
function getWebWorker(url, baseURL, workerData) {
    const workerOptions = { type: "module" };
    let scriptUrl, worker;
    if (typeof url == FUNCTION_TYPE) {
        url = url();
    }
    try {
        scriptUrl = new URL(url, baseURL);
    }
    catch (_error) {
        scriptUrl = url;
    }
    if (classicWorkersSupported) {
        try {
            worker = new Worker(scriptUrl);
        }
        catch (_error) {
            classicWorkersSupported = false;
            worker = new Worker(scriptUrl, workerOptions);
        }
    }
    else {
        worker = new Worker(scriptUrl, workerOptions);
    }
    worker.addEventListener(MESSAGE_EVENT_TYPE, event => onMessage(event, workerData));
    return worker;
}
function sendMessage(message, { worker, writer, onTaskFinished, transferStreams }) {
    try {
        let { value, readable, writable } = message;
        const transferables = [];
        if (value) {
            const { buffer, length } = value;
            if (length != buffer.byteLength) {
                value = new Uint8Array(value);
            }
            message.value = value.buffer;
            transferables.push(message.value);
        }
        if (transferStreams && transferStreamsSupported) {
            if (readable) {
                transferables.push(readable);
            }
            if (writable) {
                transferables.push(writable);
            }
        }
        else {
            message.readable = message.writable = null;
        }
        if (transferables.length) {
            try {
                worker.postMessage(message, transferables);
                return true;
            }
            catch (_error) {
                transferStreamsSupported = false;
                message.readable = message.writable = null;
                worker.postMessage(message);
            }
        }
        else {
            worker.postMessage(message);
        }
    }
    catch (error) {
        if (writer) {
            writer.releaseLock();
        }
        onTaskFinished();
        throw error;
    }
}
async function onMessage({ data }, workerData) {
    const { type, value, messageId, result, error } = data;
    const { reader, writer, resolveResult, rejectResult, onTaskFinished } = workerData;
    try {
        if (error) {
            const { message, stack, code, name } = error;
            const responseError = new Error(message);
            Object.assign(responseError, { stack, code, name });
            close(responseError);
        }
        else {
            if (type == MESSAGE_PULL) {
                const { value, done } = await reader.read();
                sendMessage({ type: MESSAGE_DATA, value, done, messageId }, workerData);
            }
            if (type == MESSAGE_DATA) {
                await writer.ready;
                await writer.write(new Uint8Array(value));
                sendMessage({ type: MESSAGE_ACK_DATA, messageId }, workerData);
            }
            if (type == MESSAGE_CLOSE) {
                close(null, result);
            }
        }
    }
    catch (error) {
        close(error);
    }
    function close(error, result) {
        if (error) {
            rejectResult(error);
        }
        else {
            resolveResult(result);
        }
        if (writer) {
            writer.releaseLock();
        }
        onTaskFinished();
    }
}
